c7n/resources/s3.py

```diff
@@ -1372,6 +1372,153 @@ def process_bucket(self, bucket):
                 return {'Name': bucket['Name'], 'State': 'ReplicationConfigUpdated'}
 
 
+@filters.register('check-public-block')
+class FilterPublicBlock(Filter):
+    """Filter for s3 bucket public blocks
+
+    :example:
+
+    .. code-block:: yaml
+
+            policies:
+              - name: CheckForPublicAclBlock-Off
+                resource: s3
+                region: us-east-1
+                filters:
+                  - type: check-public-block
+                    scope: Any
+                    state: absent
+    """
+
+    schema = type_schema(
+        'check-public-block',
+        scope={
+            'type': 'string',
+            'enum': ['BlockPublicAcls', 'IgnorePublicAcls',
+                'BlockPublicPolicy', 'RestrictPublicBuckets', 'All', 'Any']},
+        state={
+            'type': 'string',
+            'enum': ['absent', 'present']},
+        required=['scope', 'state'])
+    permissions = ("s3:GetBucketPublicAccessBlock",)
+
+    def process(self, buckets, event=None):
+        results = list()
+        with self.executor_factory(max_workers=2) as w:
+            futures = {w.submit(self.process_bucket, bucket): bucket for bucket in buckets}
+            for future in as_completed(futures):
+                if future.exception():
```

Comment:
umm, this part needs addressing. ie this is also just effectively swallowing exceptions instead of stopping the policy execution.

Reply:
so across the futures boundary, if I raise an exception, is that kind of like throwing it in Java? Where I would still need to handle it in the calling class?

Reply:
addressed in 360280b1744a653c9fd7d0bbd0ddf81b07e8182d. I think I answered my own question when saying it out loud but curious to hear your thoughts on it if you have the time.

Reply:
so there's some differences here between python 2 and 3, previously w/ py2 across thread boundaries the stack would be lost, w/ py3.3+ stack is associated to the exception.  so 
```
if future.exception():
   raise future.exception() from None
```
should retain original context/stack trace for the exception.

Reply:
Awesome! Thanks for explaining. I ended up going with sort of a raise in future, catch after execution approach.

`               
 if future.exception():
             errors.append("Message: %s Bucket: %s" % (future.exception(), bucket['Name']))
`

https://github.com/cloud-custodian/cloud-custodian/commit/360280b1744a653c9fd7d0bbd0ddf81b07e8182d

Reply:
yeah.. I know we just did this but elsewhere, but really we should be preserving the stack trace and early exiting per the sample, else we're really losing context from how/where in the thread worker the error is happening outside of guesstimating on an error message. fwiw, its something I think we need to update through out the code base around thread pool usage, now that we've dropped python 2 support.

Reply:
Ah, [PEP 409](https://www.python.org/dev/peps/pep-0409/) let me see what I can do

Reply:
as an example compare the output in

```python

from concurrent.futures import ThreadPoolExecutor, as_completed
import traceback

def foo(r):
    raise SyntaxError(str(r))

def bar(r):
    raise SyntaxError(str(r))

def run(func):
    with ThreadPoolExecutor(max_workers=3) as w:
        futures = []
        errors = []
        for r in range(3):
            futures.append(w.submit(func, r))
        for f in as_completed(futures):

            if func.__name__ == 'bar' and f.exception():
                raise f.exception()
            elif func.__name__ == 'foo' and f.exception():
                errors.append(str(f.exception()))
        raise SyntaxError("\n".join(map(str, errors)))

def main():

    try:
       run(foo)
    except Exception:
       traceback.print_exc()

    try:
       run(bar)
    except Exception:
       traceback.print_exc()

if __name__ == '__main__':
    main()
```

Reply:
> Ah, [PEP 409](https://www.python.org/dev/peps/pep-0409/) let me see what I can do

we probably don't need pep 409 per se which is casting an exception, unless we're capturing an original exception, we don't have one from the from context per se since its being stringed into a list, if we're constructing a new one, but even that its a question since we're just picking a ransoming one from a set and assuming its representative. in general just stopping on the first and exiting with the original context is probably the best option. on a failed policy, things are in a inconsistent state, if its transient re-running will resolve, if its persistent, refusing to process other resources is probably a good thing. reasoning wise on a failed policy the only rationale thing we can do is stop, and the only thing a user can assume is this set of resources matched the policy and has indeterminate state wrt to action evaluation.

Reply:
Alright, awesome! I love this. The output is much clearer.

Reply:
I agree with this approach, it is giving clearer debug logging and halting the execution does seem to be the right thing to do as soon as an error bubbles up.

Reply:
Does 077b72a78138b4fea74145329fcd289099fc4335 hit the nail on the head? I made similar updates to the action where I think this kind of 'fail early' approach also seems important.

Reply:
I also added an issue for looking at other opportunities to do this. Please feel free to assign me and I will work on it in my free time.

Reply:
it does, but there's another exception handling further that down thats doing a cast to Exception instead of just doing an argless `raise`

Reply:
Addressed in eae58596e6a20df36135aae2bd40e99a621b3881 for the argless raise. It was late 🤦‍♂️

Reply:
all good, thanks for working through all the review feedback.